worker/worker.go

```diff
@@ -570,6 +574,84 @@ func (w *worker) rhpBroadcastHandler(jc jape.Context) {
 	}
 }
 
+func (w *worker) rhpPruneContractHandlerPOST(jc jape.Context) {
+	// decode fcid
+	var id types.FileContractID
+	if jc.DecodeParam("id", &id) != nil {
+		return
+	}
+
+	// check if there's prunable data for the contract
+	ctx := jc.Request.Context()
+	n, err := w.bus.PrunableDataForContract(ctx, id)
+	if errors.Is(err, api.ErrContractNotFound) {
+		jc.Error(err, http.StatusNotFound)
+		return
+	} else if jc.Check("couldn't fetch prunable data for contract", err) != nil {
+		return
+	} else if n == 0 {
+		jc.Encode(n)
+		return
+	}
+
+	// attach gouging checker
+	gp, err := w.bus.GougingParams(ctx)
+	if jc.Check("could not get gouging parameters", err) != nil {
+		return
+	}
+	ctx = WithGougingChecker(ctx, w.bus, gp)
+
+	// fetch the contract from the bus
+	c, err := w.bus.Contract(ctx, id)
+	if errors.Is(err, api.ErrContractNotFound) {
+		jc.Error(err, http.StatusNotFound)
+		return
+	} else if jc.Check("couldn't fetch contract", err) != nil {
+		return
+	}
+
+	// convenience variables
+	hk := c.HostKey
+	hostIP := c.HostIP
+	hostAddr := c.SiamuxAddr
+
+	// prune sectors from contracts
+	var toDelete []uint64
+	if jc.Check("couldn't prune contract", w.withRevision(ctx, defaultRevisionFetchTimeout, id, hk, hostAddr, lockingPriorityPruning, 0, func(rev types.FileContractRevision) error {
+		// fetch the roots from the host
+		renterKey := w.deriveRenterKey(c.HostKey)
+		got, err := w.FetchContractRoots(ctx, hostIP, hk, renterKey, id, rev.RevisionNumber, time.Minute)
+		if err != nil {
+			return err
+		}
+
+		// fetch the roots from the bus
+		roots, err := w.bus.ContractRoots(ctx, id)
```

Comment:
I think we have a pretty severe race condition here.

When uploading a file, we update the contracts with the hosts we upload to. Only if all roots are uploaded, we update the state in the database. So as long as an upload is active, we have roots in contracts which don't exist in the database yet. Potentially causing us to prune sectors from contracts during that time.

Tbh. I haven't come up with a satisfying solution yet but the goal would be to make sure no ongoing uploads use a contract when pruning it.

Reply:
Hm damn it I figured locking the revision as well as fetching the roots from the database first would cover all races but you are right, the pruning might acquire a revision lock on an ongoing upload so there will be sectors in those contracts that are not in the database yet.

I think a reasonable solution is to put a time constraint.

Reply:
adding a pending sectors cache in the bus